File: 2.add-two-numbers/2.add-two-numbers.go
/*
 * @lc app=leetcode id=2 lang=golang
 *
 * [2] Add Two Numbers
 */
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var root, parent *ListNode
	carry := false
	for l1 != nil || l2 != nil {
		node := new(ListNode)
		if parent != nil {
			parent.Next = node
			parent = node
		}
		if root == nil {
			root = node
			parent = node
		}
		var sum int
		if l1 != nil {
			sum += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			sum += l2.Val
			l2 = l2.Next
		}
		if carry {
			sum++
			carry = false
		}
		if sum >= 10 {
			sum = sum - 10
			carry = true
		}
		node.Val = sum
	}
	if carry {
		parent.Next = &ListNode{Val: 1, Next: nil}
	}
	return root
}

File: 0002.add-two-numbers/2.add-two-numbers.py
#
# @lc app=leetcode.cn id=2 lang=python3
#
# [2] add tow numbers
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        r, n, n1, n2 = None, None, l1, l2
        overflow = False
        while n1 != None or n2 != None or overflow:
            val = 0
            if n1 != None:
                val += n1.val
                n1 = n1.next
            if n2 != None:
                val += n2.val
                n2 = n2.next
            if overflow:
                val += 1
                overflow = False
            if val >=10:
                val -= 10
                overflow = True
            t = ListNode(val)
            if r == None:
                n = r = t
            else:
                n.next = t
                n = t
        return r
        
# @lc code=end

